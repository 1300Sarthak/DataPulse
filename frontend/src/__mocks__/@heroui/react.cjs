/* eslint-disable no-undef */
const React = require('react');

module.exports = {
  Button: ({ children, ...props }) => (
    React.createElement('button', { ...props, 'data-testid': 'heroui-button' }, children)
  ),
  Spinner: (props) => (
    React.createElement('div', { ...props, 'data-testid': 'heroui-spinner' }, 'Loading...')
  ),
  Switch: (props) => (
    React.createElement('input', { type: 'checkbox', ...props, 'data-testid': 'heroui-switch' })
  ),
  Dropdown: ({ children, ...props }) => (
    React.createElement('div', { ...props, 'data-testid': 'heroui-dropdown' }, children)
  ),
  DropdownTrigger: ({ children, ...props }) => (
    React.createElement('div', { ...props, 'data-testid': 'heroui-dropdown-trigger' }, children)
  ),
  DropdownMenu: ({ children, ...props }) => (
    React.createElement('div', { ...props, 'data-testid': 'heroui-dropdown-menu' }, children)
  ),
  DropdownItem: ({ children, ...props }) => (
    React.createElement('div', { ...props, 'data-testid': 'heroui-dropdown-item' }, children)
  ),
  Card: ({ children, ...props }) => (
    React.createElement('div', { ...props, 'data-testid': 'heroui-card' }, children)
  ),
  CardBody: ({ children, ...props }) => (
    React.createElement('div', { ...props, 'data-testid': 'heroui-card-body' }, children)
  ),
  Chip: ({ children, ...props }) => (
    React.createElement('span', { ...props, 'data-testid': 'heroui-chip' }, children)
  ),
  Link: ({ children, ...props }) => (
    React.createElement('a', { ...props, 'data-testid': 'heroui-link' }, children)
  ),
  Skeleton: (props) => (
    React.createElement('div', { ...props, 'data-testid': 'heroui-skeleton' }, 'Loading...')
  ),
  Divider: (props) => React.createElement('hr', { ...props, 'data-testid': 'heroui-divider' }),
  Badge: ({ children, ...props }) => (
    React.createElement('span', { ...props, 'data-testid': 'heroui-badge' }, children)
  ),
}; 