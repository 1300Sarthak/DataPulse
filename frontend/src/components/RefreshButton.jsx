import React, { useState, useEffect, useRef } from "react";
import { Button, Spinner, Switch, Dropdown, DropdownTrigger, DropdownMenu, DropdownItem } from "@heroui/react";
import { useApiService } from '../services/api';

// Error Boundary for RefreshButton
class RefreshErrorBoundary extends React.Component {
  constructor(props) {
    super(props);
    this.state = { hasError: false, error: null };
  }

  static getDerivedStateFromError(error) {
    return { hasError: true, error };
  }

  componentDidCatch(error, errorInfo) {
    console.error("RefreshButton Error:", error, errorInfo);
  }

  render() {
    if (this.state.hasError) {
      return (
        <div className="text-center p-2">
          <div className="text-red-500 text-sm mb-1">
            <svg className="w-4 h-4 mx-auto mb-1" fill="none" stroke="currentColor" viewBox="0 0 24 24">
              <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M12 9v2m0 4h.01m-6.938 4h13.856c1.54 0 2.502-1.667 1.732-2.5L13.732 4c-.77-.833-1.964-.833-2.732 0L3.732 16.5c-.77.833.192 2.5 1.732 2.5z" />
            </svg>
            Refresh failed
          </div>
          <button
            onClick={() => this.setState({ hasError: false, error: null })}
            className="text-blue-500 hover:text-blue-700 text-xs underline"
          >
            Try again
          </button>
        </div>
      );
    }

    return this.props.children;
  }
}

const RefreshButton = ({ onRefresh, className = "" }) => {
  const apiService = useApiService();
  const [isLoading, setIsLoading] = useState(false);
  const [isDisabled, setIsDisabled] = useState(false);
  const [error, setError] = useState(null);
  const [countdown, setCountdown] = useState(0);
  const [autoRefresh, setAutoRefresh] = useState(false);
  const [autoRefreshInterval, setAutoRefreshInterval] = useState(30); // seconds
  const [nextAutoRefresh, setNextAutoRefresh] = useState(0);
  
  const intervalRef = useRef(null);
  const autoRefreshCountdownRef = useRef(null);

  const intervals = [
    { value: 15, label: "15s" },
    { value: 30, label: "30s" },
    { value: 60, label: "1m" },
    { value: 300, label: "5m" },
  ];

  // Countdown timer for disabled state
  useEffect(() => {
    let interval;
    if (countdown > 0) {
      interval = setInterval(() => {
        setCountdown(prev => {
          if (prev <= 1) {
            setIsDisabled(false);
            return 0;
          }
          return prev - 1;
        });
      }, 1000);
    }
    return () => clearInterval(interval);
  }, [countdown]);

  // Auto-refresh countdown timer
  useEffect(() => {
    if (autoRefresh && nextAutoRefresh > 0) {
      autoRefreshCountdownRef.current = setInterval(() => {
        setNextAutoRefresh(prev => {
          if (prev <= 1) {
            handleRefresh();
            return autoRefreshInterval;
          }
          return prev - 1;
        });
      }, 1000);
    }
    return () => {
      if (autoRefreshCountdownRef.current) {
        clearInterval(autoRefreshCountdownRef.current);
      }
    };
  }, [autoRefresh, autoRefreshInterval]);

  // Auto-refresh interval
  useEffect(() => {
    if (autoRefresh) {
      intervalRef.current = setInterval(() => {
        handleRefresh();
      }, autoRefreshInterval * 1000);
      
      setNextAutoRefresh(autoRefreshInterval);
    } else {
      if (intervalRef.current) {
        clearInterval(intervalRef.current);
        intervalRef.current = null;
      }
      setNextAutoRefresh(0);
    }

    return () => {
      if (intervalRef.current) {
        clearInterval(intervalRef.current);
      }
    };
  }, [autoRefresh, autoRefreshInterval]);

  const handleRefresh = async () => {
    if (isLoading || isDisabled) return;

    setIsLoading(true);
    setError(null);

    try {
      // Call the refresh endpoint
      const response = await apiService.refresh();

      if (!response.ok) {
        throw new Error(`Refresh failed: ${response.status}`);
      }

      // Trigger re-render of all components
      if (onRefresh) {
        onRefresh();
      }

      // Disable button for 5 seconds (only for manual refresh)
      if (!autoRefresh) {
        setIsDisabled(true);
        setCountdown(5);
      }

    } catch (err) {
      setError(err.message);
      console.error("Refresh error:", err);
    } finally {
      setIsLoading(false);
    }
  };

  const handleAutoRefreshToggle = () => {
    setAutoRefresh(!autoRefresh);
    if (!autoRefresh) {
      // Starting auto-refresh, trigger first refresh immediately
      handleRefresh();
    }
  };

  const handleIntervalChange = (newInterval) => {
    setAutoRefreshInterval(newInterval);
    if (autoRefresh) {
      setNextAutoRefresh(newInterval);
    }
  };

  const getButtonText = () => {
    if (isLoading) return "Refreshing...";
    if (isDisabled) return `Refresh (${countdown}s)`;
    if (autoRefresh && nextAutoRefresh > 0) return `Auto (${nextAutoRefresh}s)`;
    return "Refresh";
  };

  const getButtonIcon = () => {
    if (isLoading) {
      return <Spinner size="sm" color="current" />;
    }
    if (autoRefresh) {
      return (
        <svg className="w-4 h-4" fill="none" stroke="currentColor" viewBox="0 0 24 24">
          <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M13 10V3L4 14h7v7l9-11h-7z" />
        </svg>
      );
    }
    return (
      <svg className="w-4 h-4" fill="none" stroke="currentColor" viewBox="0 0 24 24">
        <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M4 4v5h.582m15.356 2A8.001 8.001 0 004.582 9m0 0H9m11 11v-5h-.581m0 0a8.003 8.003 0 01-15.357-2m15.357 2H15" />
      </svg>
    );
  };

  if (error) {
    return (
      <div className={`text-center ${className}`}>
        <div className="text-red-500 text-sm mb-2">
          <svg className="w-4 h-4 mx-auto mb-1" fill="none" stroke="currentColor" viewBox="0 0 24 24">
            <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M12 8v4m0 4h.01M21 12a9 9 0 11-18 0 9 9 0 0118 0z" />
          </svg>
          {error}
        </div>
        <Button
          size="sm"
          color="primary"
          variant="flat"
          onClick={handleRefresh}
          disabled={isLoading || isDisabled}
        >
          {getButtonIcon()}
          Retry Refresh
        </Button>
      </div>
    );
  }

  return (
    <div className={`flex items-center gap-2 ${className}`}>
      {/* Auto-refresh toggle */}
      <div className="flex items-center gap-1">
        <span className="text-xs text-gray-500">Auto</span>
        <Switch
          size="sm"
          isSelected={autoRefresh}
          onValueChange={handleAutoRefreshToggle}
          color="primary"
        />
      </div>

      {/* Interval selector */}
      {autoRefresh && (
        <Dropdown>
          <DropdownTrigger>
            <Button size="sm" variant="bordered" className="min-w-0 px-2">
              {intervals.find(i => i.value === autoRefreshInterval)?.label || "30s"}
            </Button>
          </DropdownTrigger>
          <DropdownMenu
            aria-label="Auto-refresh intervals"
            onAction={(key) => handleIntervalChange(Number(key))}
          >
            {intervals.map((interval) => (
              <DropdownItem key={interval.value}>
                {interval.label}
              </DropdownItem>
            ))}
          </DropdownMenu>
        </Dropdown>
      )}

      {/* Manual refresh button */}
      <Button
        size="sm"
        color={autoRefresh ? "success" : "primary"}
        variant="flat"
        onClick={handleRefresh}
        disabled={isLoading || isDisabled}
        startContent={getButtonIcon()}
      >
        {getButtonText()}
      </Button>
    </div>
  );
};

// Wrapped component with error boundary
const RefreshButtonWithErrorBoundary = (props) => (
  <RefreshErrorBoundary>
    <RefreshButton {...props} />
  </RefreshErrorBoundary>
);

export default RefreshButtonWithErrorBoundary; 