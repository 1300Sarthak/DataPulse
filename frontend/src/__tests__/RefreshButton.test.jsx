/* global describe, it, expect, beforeEach, afterEach, jest, global */
import React from "react";
import { render, screen, fireEvent, waitFor, act } from "@testing-library/react";
import "@testing-library/jest-dom";
import RefreshButton from "../components/RefreshButton";

// Mock fetch globally
global.fetch = jest.fn();

// Mock timers
jest.useFakeTimers();

describe("RefreshButton", () => {
  beforeEach(() => {
    fetch.mockClear();
    jest.clearAllTimers();
  });

  afterEach(() => {
    jest.runOnlyPendingTimers();
    jest.useRealTimers();
  });

  it("renders refresh button with default state", () => {
    render(<RefreshButton onRefresh={jest.fn()} />);
    
    expect(screen.getByText("Refresh")).toBeInTheDocument();
    expect(screen.getByText("Auto")).toBeInTheDocument();
    expect(screen.getByRole("switch")).toBeInTheDocument();
  });

  it("calls refresh endpoint and onRefresh callback on button click", async () => {
    const mockOnRefresh = jest.fn();
    fetch.mockResolvedValueOnce({ ok: true });

    render(<RefreshButton onRefresh={mockOnRefresh} />);
    
    const refreshButton = screen.getByText("Refresh");
    fireEvent.click(refreshButton);

    await waitFor(() => {
      expect(fetch).toHaveBeenCalledWith("/refresh", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
      });
    });

    await waitFor(() => {
      expect(mockOnRefresh).toHaveBeenCalled();
    });
  });

  it("shows loading state during refresh", async () => {
    fetch.mockImplementation(() => new Promise(resolve => setTimeout(() => resolve({ ok: true }), 100)));

    render(<RefreshButton onRefresh={jest.fn()} />);
    
    const refreshButton = screen.getByText("Refresh");
    fireEvent.click(refreshButton);

    expect(screen.getByText("Refreshing...")).toBeInTheDocument();
  });

  it("disables button for 5 seconds after manual refresh", async () => {
    fetch.mockResolvedValueOnce({ ok: true });

    render(<RefreshButton onRefresh={jest.fn()} />);
    
    const refreshButton = screen.getByText("Refresh");
    fireEvent.click(refreshButton);

    await waitFor(() => {
      expect(screen.getByText("Refresh (5s)")).toBeInTheDocument();
    });

    // Fast-forward 4 seconds
    act(() => {
      jest.advanceTimersByTime(4000);
    });

    expect(screen.getByText("Refresh (1s)")).toBeInTheDocument();

    // Fast-forward 1 more second
    act(() => {
      jest.advanceTimersByTime(1000);
    });

    expect(screen.getByText("Refresh")).toBeInTheDocument();
  });

  it("handles refresh errors gracefully", async () => {
    fetch.mockRejectedValueOnce(new Error("Network error"));

    render(<RefreshButton onRefresh={jest.fn()} />);
    
    const refreshButton = screen.getByText("Refresh");
    fireEvent.click(refreshButton);

    await waitFor(() => {
      expect(screen.getByText("Network error")).toBeInTheDocument();
    });

    expect(screen.getByText("Retry Refresh")).toBeInTheDocument();
  });

  it("handles HTTP error responses", async () => {
    fetch.mockResolvedValueOnce({ ok: false, status: 500 });

    render(<RefreshButton onRefresh={jest.fn()} />);
    
    const refreshButton = screen.getByText("Refresh");
    fireEvent.click(refreshButton);

    await waitFor(() => {
      expect(screen.getByText("Refresh failed: 500")).toBeInTheDocument();
    });
  });

  it("toggles auto-refresh when switch is clicked", () => {
    render(<RefreshButton onRefresh={jest.fn()} />);
    
    const autoSwitch = screen.getByRole("switch");
    expect(autoSwitch).not.toBeChecked();

    fireEvent.click(autoSwitch);
    expect(autoSwitch).toBeChecked();

    fireEvent.click(autoSwitch);
    expect(autoSwitch).not.toBeChecked();
  });

  it("shows interval selector when auto-refresh is enabled", () => {
    render(<RefreshButton onRefresh={jest.fn()} />);
    
    const autoSwitch = screen.getByRole("switch");
    fireEvent.click(autoSwitch);

    expect(screen.getByText("30s")).toBeInTheDocument();
  });

  it("hides interval selector when auto-refresh is disabled", () => {
    render(<RefreshButton onRefresh={jest.fn()} />);
    
    const autoSwitch = screen.getByRole("switch");
    fireEvent.click(autoSwitch);
    expect(screen.getByText("30s")).toBeInTheDocument();

    fireEvent.click(autoSwitch);
    expect(screen.queryByText("30s")).not.toBeInTheDocument();
  });

  it("starts auto-refresh immediately when enabled", async () => {
    fetch.mockResolvedValue({ ok: true });
    const mockOnRefresh = jest.fn();

    render(<RefreshButton onRefresh={mockOnRefresh} />);
    
    const autoSwitch = screen.getByRole("switch");
    fireEvent.click(autoSwitch);

    await waitFor(() => {
      expect(fetch).toHaveBeenCalled();
      expect(mockOnRefresh).toHaveBeenCalled();
    });
  });

  it("changes button color when auto-refresh is active", () => {
    render(<RefreshButton onRefresh={jest.fn()} />);
    
    const autoSwitch = screen.getByRole("switch");
    fireEvent.click(autoSwitch);

    const refreshButton = screen.getByText("Refresh");
    expect(refreshButton.closest('button')).toHaveClass('bg-success');
  });

  it("does not disable button during auto-refresh", async () => {
    fetch.mockResolvedValue({ ok: true });

    render(<RefreshButton onRefresh={jest.fn()} />);
    
    const autoSwitch = screen.getByRole("switch");
    fireEvent.click(autoSwitch);

    await waitFor(() => {
      expect(fetch).toHaveBeenCalled();
    });

    // Button should not be disabled during auto-refresh
    const refreshButton = screen.getByText("Refresh");
    expect(refreshButton.closest('button')).not.toBeDisabled();
  });

  it("cleans up intervals when component unmounts", () => {
    const { unmount } = render(<RefreshButton onRefresh={jest.fn()} />);
    
    const autoSwitch = screen.getByRole("switch");
    fireEvent.click(autoSwitch);

    // Should not throw when unmounting with active intervals
    expect(() => unmount()).not.toThrow();
  });

  it("handles error boundary gracefully", () => {
    // Mock console.error to avoid noise in tests
    const originalError = console.error;
    console.error = jest.fn();

    // Create a component that throws an error
    const ErrorComponent = () => {
      throw new Error("Test error");
    };

    // This should not crash the app
    expect(() => render(<ErrorComponent />)).toThrow();

    console.error = originalError;
  });

  it("matches snapshot", () => {
    const { container } = render(<RefreshButton onRefresh={jest.fn()} />);
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with auto-refresh enabled", () => {
    const { container } = render(<RefreshButton onRefresh={jest.fn()} />);
    
    const autoSwitch = screen.getByRole("switch");
    fireEvent.click(autoSwitch);

    expect(container).toMatchSnapshot();
  });
}); 