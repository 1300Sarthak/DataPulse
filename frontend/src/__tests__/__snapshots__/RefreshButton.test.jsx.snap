// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`RefreshButton matches snapshot 1`] = `
<div>
  <div
    class="flex items-center gap-2 "
  >
    <div
      class="flex items-center gap-1"
    >
      <span
        class="text-xs text-gray-500"
      >
        Auto
      </span>
      <label
        class="group relative max-w-fit inline-flex items-center justify-start cursor-pointer touch-none tap-highlight-transparent select-none"
      >
        <input
          aria-labelledby="«r2m»"
          class="font-inherit text-[100%] leading-[1.15] m-0 p-0 overflow-visible box-border absolute top-0 w-full h-full opacity-[0.0001] z-[1] cursor-pointer disabled:cursor-default"
          data-react-aria-pressable="true"
          role="switch"
          tabindex="0"
          type="checkbox"
          value=""
        />
        <span
          aria-hidden="true"
          class="px-1 relative inline-flex items-center justify-start flex-shrink-0 overflow-hidden bg-default-200 rounded-full outline-none group-data-[focus-visible=true]:z-10 group-data-[focus-visible=true]:ring-2 group-data-[focus-visible=true]:ring-focus group-data-[focus-visible=true]:ring-offset-2 group-data-[focus-visible=true]:ring-offset-background group-data-[selected=true]:bg-primary group-data-[selected=true]:text-primary-foreground w-10 h-6 transition-background"
        >
          <span
            class="z-10 flex items-center justify-center bg-white shadow-small rounded-full origin-right pointer-events-none w-4 h-4 text-tiny group-data-[selected=true]:ms-4 transition-all group-data-[pressed=true]:w-5 group-data-[selected]:group-data-[pressed]:ml-3"
          />
        </span>
      </label>
    </div>
    <button
      class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-3 min-w-16 h-8 text-tiny gap-2 rounded-small [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-primary/20 text-primary-600 data-[hover=true]:opacity-hover"
      data-react-aria-pressable="true"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="w-4 h-4"
        fill="none"
        focusable="false"
        stroke="currentColor"
        viewBox="0 0 24 24"
      >
        <path
          d="M4 4v5h.582m15.356 2A8.001 8.001 0 004.582 9m0 0H9m11 11v-5h-.581m0 0a8.003 8.003 0 01-15.357-2m15.357 2H15"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
      Refresh
    </button>
  </div>
</div>
`;

exports[`RefreshButton matches snapshot with auto-refresh enabled 1`] = `
<div>
  <div
    class="flex items-center gap-2 "
  >
    <div
      class="flex items-center gap-1"
    >
      <span
        class="text-xs text-gray-500"
      >
        Auto
      </span>
      <label
        class="group relative max-w-fit inline-flex items-center justify-start cursor-pointer touch-none tap-highlight-transparent select-none"
        data-selected="true"
      >
        <input
          aria-labelledby="«r2n»"
          class="font-inherit text-[100%] leading-[1.15] m-0 p-0 overflow-visible box-border absolute top-0 w-full h-full opacity-[0.0001] z-[1] cursor-pointer disabled:cursor-default"
          data-react-aria-pressable="true"
          role="switch"
          tabindex="0"
          type="checkbox"
          value=""
        />
        <span
          aria-hidden="true"
          class="px-1 relative inline-flex items-center justify-start flex-shrink-0 overflow-hidden bg-default-200 rounded-full outline-none group-data-[focus-visible=true]:z-10 group-data-[focus-visible=true]:ring-2 group-data-[focus-visible=true]:ring-focus group-data-[focus-visible=true]:ring-offset-2 group-data-[focus-visible=true]:ring-offset-background group-data-[selected=true]:bg-primary group-data-[selected=true]:text-primary-foreground w-10 h-6 transition-background"
        >
          <span
            class="z-10 flex items-center justify-center bg-white shadow-small rounded-full origin-right pointer-events-none w-4 h-4 text-tiny group-data-[selected=true]:ms-4 transition-all group-data-[pressed=true]:w-5 group-data-[selected]:group-data-[pressed]:ml-3"
          />
        </span>
      </label>
    </div>
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 border-medium h-8 text-tiny gap-2 rounded-small [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-transparent border-default text-foreground data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased min-w-0 px-2"
      data-react-aria-pressable="true"
      data-slot="trigger"
      id="react-aria-«r2o»"
      tabindex="0"
      type="button"
    >
      30s
    </button>
    <button
      class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-3 min-w-16 h-8 text-tiny gap-2 rounded-small [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-success/20 text-success-700 dark:text-success data-[hover=true]:opacity-hover"
      data-react-aria-pressable="true"
      disabled=""
      tabindex="0"
      type="button"
    >
      <div
        aria-hidden="true"
        aria-label="Loading"
        class="relative inline-flex flex-col gap-2 items-center justify-center"
        focusable="false"
      >
        <div
          class="relative flex w-5 h-5"
        >
          <i
            class="absolute w-full h-full rounded-full border-2 border-b-current animate-spinner-ease-spin border-solid border-t-transparent border-l-transparent border-r-transparent"
          />
          <i
            class="absolute w-full h-full rounded-full border-2 border-b-current opacity-75 animate-spinner-linear-spin border-dotted border-t-transparent border-l-transparent border-r-transparent"
          />
        </div>
      </div>
      Refreshing...
    </button>
  </div>
</div>
`;
